design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,414 @@
+  prefix: 24
+  preAllocations:
+    "RenderedData-10": 192.168.0.9
+    "RenderedData-9": 192.168.0.8
+  namePrefix: "provisioning"
+status:
+  lastUpdated: "2020-04-02T06:36:09Z"
+  allocations:
+    "RenderedData-1": "192.168.0.11"
+    "RenderedData-10": 192.168.0.9
+    "RenderedData-9": 192.168.0.8
+```
+
+In the *spec* of *IPPool*, there would be a *pools* list, that would
+contain a list of IP address pools, with the *start* and *end* attributes
+giving the start and end ip addresses of the pool. The *subnet* field allows to
+verify that the allocated IP is in the pool and from which the start and end ip
+addresses can be inferred. Specifying single ip addresses can be achieved by
+setting the start and end ip address to that single ip address.
+
+The *prefix* and *gateway* parameters can be given for each pool of the list,
+or globally. If they are given for a pool they will override the global
+settings, that are default values. The *prefix* and *gateway* will be set on
+the *IPAddress* and can be fetched from a *Template*.
+
+The *allocations* fields is a map of object name and ip address that allow a
+user to specify a set of static allocations for some objects.
+
+The *namePrefix* contains the prefix used to name the IPAddress objects
+created. It must remain the same for a subnet, across updates or changes in the
+IPPool object to keep the existing leases.
+
+The *status* would contain a *lastUpdated* field with the timestamp of the last
+update. In case of an error during the allocation (pool exhaustion for example),
+the error would be reported on the Claim object, in the *errorMessage* field.
+The *allocations* map will map the IP address to the *RenderedData* object it was
+allocated for and the *addresses* will map the *RenderedData* objects with the
+*IPAddress* objects.
+
+The *IPAddress* object would be the following
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPAddress
+metadata:
+  name: pool-1-192-168-0-11
+  namespace: default
+  ownerReferences:
+  - apiVersion: ipam.metal3.io/v1alpha1
+    kind: IPPool
+    name: pool-1
+  - apiVersion: ipam.metal3.io/v1alpha1
+    kind: IPClaim
+    name: RenderedData-1
+spec:
+  claim:
+    Name: RenderedData-1
+  Address: 192.168.0.11
+  prefix: 24
+  gateway: 192.168.0.1
+  pool:
+    Name: pool-1
+status:
+  ready: true
+```
+
+The *IPClaim* object would be the following
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPClaim
+metadata:
+  name: RenderedData-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: metadata.metal3.io/v1alpha1
+    kind: Data
+    name: data-1
+spec:
+  owner:
+    name: RenderedData-1
+  pool:
+    name: pool-1
+status:
+  ipAddress:
+    name: pool-1-192-168-0-11
+  errorMessage: ""
+```
+
+For each *IPClaim*, the controller reconciling the *IPClaim* will select an
+available IP address randomly from the *IPPool*, if the object name is not in
+the *allocations* map in the object *spec* or in the *addresses* map in the
```

Comment:
This statement about using the addresses map is in conflict with the statement on line 280 about not using the status field. I think this line is left over from before the edits made further down?

Reply:
yes, definitely. I implemented it based only on the list of objects, not on the status